### PR TITLE
Enable packed sequence in SFT with Chat dataset

### PR DESCRIPTION
**Pull request by jenchen13**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Enable packed sequence in SFT with Chat dataset

**Collection**: llm

# Changelog

- Enable packed sequence in ChatDataModule
- add packed sequence flag to `gpt_train.py` script for SFT

# Testing

- Tested SFT on Qwen 3

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/utils/sequence_packing_utils.py [hunk 1]
```diff
@@ -118,25 +118,20 @@ def create_hist(dataset: np.array, truncate_seq_len: int):
...
-        sequences[seq_len].append(item_dict)
-        counts[seq_len] += 1
+        if seq_len <= truncate_seq_len:
+            sequences[seq_len].append(item_dict)
+            counts[seq_len] += 1
 
     logging.debug("Histogram of sequence lengths")
     logging.debug(counts)
-
-    histogram = []
```
Comment (jenchen13) [inline nemo/utils/sequence_packing_utils.py]: this code seems redundant, so I removed it. Lmk if it's actually needed

Comment (cuichenx) [reply nemo/utils/sequence_packing_utils.py]: good call

File: scripts/llm/gpt_train.py [hunk 2]
```diff
@@ -159,21 +165,20 @@ def _read_chat_template(template_path: str):
...
         data = ChatDataModule(
             dataset_root=args.data_paths[0],
             seq_length=args.seq_length,
             tokenizer=tokenizer,
             global_batch_size=args.gbs,
             micro_batch_size=args.mbs,
             use_hf_tokenizer_chat_template=True,
+            packed_sequence_specs=ps if args.pack_sequence else None,
         )
+        data.prepare_data()  # sequence packing
```
Comment (cuichenx) [inline scripts/llm/gpt_train.py]: prepare_data shouldn't need to be called explicitly. PTL will call it on a single process

File: nemo/collections/llm/gpt/data/packed_sequence.py [hunk 3]
```diff
@@ -29,7 +29,8 @@ def tokenize_dataset(
     tokenizer: TokenizerSpec,
     max_seq_length: int,
     seed: int,
-    dataset_kwargs: Optional[dict],
+    chat: bool = False,
+    dataset_kwargs: Optional[dict] = None,
```
Comment (cuichenx) [inline nemo/collections/llm/gpt/data/packed_sequence.py:32]: could you add these to the docstring?

**Comment by ChenhanYu** [comment]
LGTM; quick questions,

1. Is micro_batch_size >1 supported with sequence packing?
2. Have we checked what special token is inserted to separate 2 different conversations? 
3. Have we checked that if the attention_mask is reset between different conversations?

File: nemo/utils/sequence_packing_utils.py (hunk 1 above)
Comment (soluwalana) [reply nemo/utils/sequence_packing_utils.py]: The histogram is needed in Customizer for some additional optimization that we do

Comment (soluwalana) [reply nemo/utils/sequence_packing_utils.py]: https://gitlab-master.nvidia.com/aire/microservices/nmp/-/blob/main/services/customizer/src/training/nemo/sequence_packing.py?ref_type=heads#L334

Comment (soluwalana) [reply nemo/utils/sequence_packing_utils.py]: Also this isn't redundant, this returns 0s for all values where the sequences are missing

File: nemo/utils/sequence_packing_utils.py [hunk 4]
```diff
@@ -118,25 +118,20 @@ def create_hist(dataset: np.array, truncate_seq_len: int):
...
     for item_dict in dataset:
         # Minus 1 here to account for the fact that transformer input and label
         # have one less token than the full sequence.
         # Input is missing the last token and label is missing the first token
         # (this way the tokens are aligned for next token prediction).
         # We want pack size to be the length of the actual input and label, hence minus 1.
         seq_len = len(item_dict["input_ids"]) - 1
-        sequences[seq_len].append(item_dict)
-        counts[seq_len] += 1
+        if seq_len <= truncate_seq_len:
```
Comment (mxinO) [inline nemo/utils/sequence_packing_utils.py:128]: It seems the truncation of chat data is done in collect_fn, that's why you can see seq_len > truncate_seq_len, for chat data here, you need correctly handle the truncation before packing.

Comment (jenchen13) [reply nemo/utils/sequence_packing_utils.py:128]: where is the collect_fn you are referring to?

**Comment by simoneangarano** [comment]
Hi, is this still in progress or is sequence packing support for ChatDataModule coming anytime soon?
